From ee9c4ad32066ca2ebb638cb52e7cf88d85c5da8f Mon Sep 17 00:00:00 2001
Message-ID: <ee9c4ad32066ca2ebb638cb52e7cf88d85c5da8f.1737803595.git.lorenzo@kernel.org>
In-Reply-To: <1e82f12b417fa132b7338df50ddd5499bcbd3f26.1737803595.git.lorenzo@kernel.org>
References: <1e82f12b417fa132b7338df50ddd5499bcbd3f26.1737803595.git.lorenzo@kernel.org>
From: Lorenzo Bianconi <lorenzo@kernel.org>
Date: Fri, 24 Jan 2025 13:32:20 +0100
Subject: [PATCH net-next 2/9] net: dsa: mt7530: Enable cpu port sptag for
 EN7581 SoC

Packet Processor Engine (PPE) module used for hw acceleration on EN7581
SoC, in order to properly parse packets, requires untagged packets.
For this reason, enable CPU port STAG for EN7581 SoC.
This is a preliminary patch to enable netfilter flowtable hw offloading
on EN7581 SoC.

Signed-off-by: Lorenzo Bianconi <lorenzo@kernel.org>
---
 drivers/net/dsa/mt7530.c | 17 ++++++++++++++++-
 drivers/net/dsa/mt7530.h |  4 ++++
 2 files changed, 20 insertions(+), 1 deletion(-)

--- a/drivers/net/dsa/mt7530.c
+++ b/drivers/net/dsa/mt7530.c
@@ -3162,6 +3162,21 @@ static int mt7988_setup(struct dsa_switc
 	return mt7531_setup_common(ds);
 }
 
+static int en7581_setup(struct dsa_switch *ds)
+{
+	struct mt7530_priv *priv = ds->priv;
+	int err;
+
+	err = mt7988_setup(ds);
+	if (err)
+		return err;
+
+	mt7530_write(priv, MT753X_CPORT_SPTAG_CFG,
+		     CPORT_SW2FE_STAG_EN | CPORT_FE2SW_STAG_EN);
+
+	return 0;
+}
+
 const struct dsa_switch_ops mt7530_switch_ops = {
 	.get_tag_protocol	= mtk_get_tag_protocol,
 	.setup			= mt753x_setup,
@@ -3250,7 +3265,7 @@ const struct mt753x_info mt753x_table[]
 	[ID_EN7581] = {
 		.id = ID_EN7581,
 		.pcs_ops = &mt7530_pcs_ops,
-		.sw_setup = mt7988_setup,
+		.sw_setup = en7581_setup,
 		.phy_read_c22 = mt7531_ind_c22_phy_read,
 		.phy_write_c22 = mt7531_ind_c22_phy_write,
 		.phy_read_c45 = mt7531_ind_c45_phy_read,
--- a/drivers/net/dsa/mt7530.h
+++ b/drivers/net/dsa/mt7530.h
@@ -615,6 +615,10 @@ enum mt7531_xtal_fsel {
 #define  MT7531_GPIO12_RG_RXD3_MASK	GENMASK(19, 16)
 #define  MT7531_EXT_P_MDIO_12		(2 << 16)
 
+#define MT753X_CPORT_SPTAG_CFG		0x7c10
+#define  CPORT_SW2FE_STAG_EN		BIT(1)
+#define  CPORT_FE2SW_STAG_EN		BIT(0)
+
 /* Registers for LED GPIO control (MT7530 only)
  * All registers follow this pattern:
  * [ 2: 0]  port 0
